lib/kernel.js

```diff
@@ -16,34 +16,89 @@ import store from "./store";
 import WatchesStore from "./store/watches";
 import OutputStore from "./store/output";
 import HydrogenKernel from "./plugin-api/hydrogen-kernel";
+import InputView from "./input-view";
+import KernelTransport from "./kernel-transport";
+import ZMQKernel from "./zmq-kernel";
+import type { ResultsCallback } from "./kernel-transport";
+
+interface KernelMiddlewareStack {
+  interrupt: () => void;
+  shutdown: () => void;
+  restart: (onRestarted: ?Function) => void;
+  execute: (
+    code: string,
+    callWatches: boolean,
+    onResults: ResultsCallback
+  ) => void;
+  complete: (code: string, onResults: ResultsCallback) => void;
+  inspect: (
+    code: string,
+    cursorPos: number,
+    onResults: ResultsCallback
+  ) => void;
+}
 
 export default class Kernel {
-  @observable executionState = "loading";
   @observable inspector = { bundle: {} };
   outputStore = new OutputStore();
 
-  kernelSpec: Kernelspec;
-  grammar: atom$Grammar;
-  language: string;
-  displayName: string;
   watchesStore: WatchesStore;
-  watchCallbacks: Array<Function> = [];
   emitter = new Emitter();
   pluginWrapper: HydrogenKernel | null = null;
+  kernel: ZMQKernel | KernelTransport; // XXX(nikita): need explicit ZMQKernel for plugins to typecheck
+  middlewareStack: KernelMiddlewareStack;
 
-  constructor(kernelSpec: Kernelspec, grammar: atom$Grammar) {
-    this.kernelSpec = kernelSpec;
-    this.grammar = grammar;
-
-    this.language = kernelSpec.language.toLowerCase();
-    this.displayName = kernelSpec.display_name;
+  constructor(kernel: KernelTransport) {
+    this.kernel = kernel;
 
     this.watchesStore = new WatchesStore(this);
+
+    this.middlewareStack = {
+      interrupt: () => this.kernel.interrupt(),
+      shutdown: () => this.kernel.shutdown(),
+      restart: (onRestarted: ?Function) => this.kernel.restart(onRestarted),
+      execute: (
+        code: string,
+        callWatches: boolean,
+        onResults: ResultsCallback
+      ) => this.kernel.execute(code, callWatches, onResults),
+      complete: (code: string, onResults: ResultsCallback) =>
+        this.kernel.complete(code, onResults),
+      inspect: (code: string, cursorPos: number, onResults: ResultsCallback) =>
+        this.kernel.inspect(code, cursorPos, onResults)
+    };
+  }
+
+  get kernelSpec(): Kernelspec {
+    return this.kernel.kernelSpec;
```

Comment:
All these methods are telling you that Kernel(=HKernel) should extend KernelTransport(=Kernel).

Reply:
But they have a different interface!

`KernelTransport` has `onResults: (message: JupyterMessageObject, channel: string) => void`, and `Kernel` has `onResults: (message: HydrogenInternalUndocumentedFormatObject) => void`

I thought about this, but it would involve a major rewrite of hydrogen to remove all use of the hydrogen-internal format for results

Reply:
Don't change the interface. This class is used by Hydrogen. If you change the interface you'll have to propagate the changes to the whole of Hydrogen.

I'm OK, if you don't want to extend `KernelTransport(=Kernel)`, but you shouldn't change the interface in this PR.

Reply:
> I thought about inheritance, but it would involve a major rewrite of hydrogen to remove all use of the hydrogen-internal format for results

I don't follow the reasoning above. Kernel(=HKernel) is used internally! If you change its interface, you'll have to propagate the changes to the whole of Hydrogen.